gh-81554: Add add_reader support to ProactorEventLoop

minrk opened the pull request:
Runs `select.select` in a background thread when `add_reader`/writer are called. Thread is terminated when the event loop is destroyed.

Solves a major problem with the default event loop lacking these methods (see #81554 for use cases).

Imports `SelectorThread` implementation to `asyncio._selector_thread` from Tornado 6.5.2 and uses it in ProactorEventLoop, fixing a longstanding compatibility problem with the default Windows event loop. Feel free to review `_selector_thread`, but note that it currently has _almost_ no modifications to the original file from tornado (only removing some type hints not supported by the standard library).

Since this is only run on demand, no thread will be spawned for existing code that works on ProactorEventLoop, only new code that previously only worked on SelectorEventLoop or tornado's AddThreadEventLoop will now work with the default event loop.

It _is_ possible to do this without a thread (`trio` does it), but it seems this is the simplest, most maintainable solution, as indicated in the discussion in #81554.

<!-- gh-issue-number: gh-81554 -->
* Issue: gh-81554
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--141834.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

minrk committed:
import tornado selector thread

v6.5.1
License: Apache-2.0
Copyright (c) 2025 The Tornado Authors

unmodified from original, so we can track changes

minrk committed:
remove some unused typing from selector thread

minrk committed:
Add ProactorEventLoop.add_reader and friends

via background SelectorThread, imported from tornado

minrk committed:
update spdx comments for tornado 6.5.2

StanFromIreland commented on Doc/library/asyncio-eventloop.rst:
```diff
@@ -1051,6 +1051,9 @@ Watching file descriptors
+.. versionchanged:: 3.15
+
+   Added support for these methods to :class:`ProactorEventLoop`.
 
```
```suggestion


```

StanFromIreland commented on Doc/library/asyncio-platforms.rst:
```diff
@@ -59,15 +59,19 @@ All event loops on Windows do not support the following methods:
-* The :meth:`loop.add_reader` and :meth:`loop.add_writer`
-  methods are not supported.
+* :meth:`loop.add_reader` and :meth:`loop.add_writer` only accept
+  socket handles (e.g. pipe file descriptors are not supported).
```
```suggestion
  socket handles (for example, pipe file descriptors are not supported).
```
[Per our style guide.](https://devguide.python.org/documentation/style-guide/#use-simple-language)

StanFromIreland commented on Doc/library/asyncio-platforms.rst:
```diff
@@ -59,15 +59,19 @@ All event loops on Windows do not support the following methods:
+.. versionadded:: 3.15
+
+   Support for :meth:`loop.add_reader`, :meth:`loop.add_writer` added to :class:`ProactorEventLoop`.
 
```
```suggestion


```
Please keep the same spacing as there was previously.

StanFromIreland commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+# Contains code from https://github.com/tornadoweb/tornado/tree/v6.5.2
+# SPDX-License-Identifier: PSF-2.0 AND Apache-2.0
+# SPDX-FileCopyrightText: Copyright (c) 2025 The Tornado Authors
```
Should this not be in https://github.com/python/cpython/blob/main/Misc/sbom.spdx.json?

See the [devguide](https://devguide.python.org/developer-workflow/sbom/#adding-a-new-dependency).

minrk replied:
This is not perfectly as-is, since it is adapted and an excerpt, like `Lib/asyncio/events.py` and a number of others. That doesn't seem to fit the SBOM requirements, in my understanding? I don't see any of the other similarly adapted files in sbom.spdx.json.

minrk replied:
After addressing review, it has diverged more substantially. This comment structure seems to be how adapted/partial content is handled. Is there anything in the dev guide for that? I couldn't find it. The SBOM page doesn't seem to apply or mention this sort of case.

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+            self._select_cond.notify()
+
+    def _run_select(self) -> None:
+        while True:
```
```suggestion
        while self._closed:
```

minrk replied:
This would never return, as `self._closed` is only set after this thread is joined. Perhaps you mean `while not self._closing_selector`?

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+                else:
+                    raise
+
+            try:
```
You may exit if self._closed is true.

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+                self._real_loop.call_soon_threadsafe(
+                    self._handle_select, rs, ws, context=self._main_thread_ctx
+                )
+            except RuntimeError:
```
I'm not comfortable with ignoring such exception. Can't you fix the race condition instead? Same remark for AttributeError.

minrk replied:
I believe this race is addressed now that this is integrated into the actual EventLoop class, rather than in a wrapper like Tornado has to do. I've removed the checks.

minrk committed:
Apply suggestions from code review

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+    Optional,
+    Protocol,
+    Set,
+    Tuple,
```
Can't you use dict, list, set, tuple types instead?

minrk replied:
yes, I didn't update anything from the upstream file that wasn't broken, but now I've removed much of the code required to support older Pythons now.

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+Adapted from Tornado 6.5.2
+"""
+
+from __future__ import annotations
```
Is it still needed on Python 3.14?

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+        self._thread_manager_handle = self._thread_manager()
+
+        async def thread_manager_anext() -> None:
+            # the anext builtin wasn't added until 3.10. We just need to iterate
```
This code is for Python 3.15.

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+        # can be shut down in this way (non-daemon threads would require the
+        # introduction of a new hook: https://bugs.python.org/issue41962)
+        self._thread = threading.Thread(
+            name="Tornado selector",
```
You may replace references to Tornado with asyncio :-)

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+        # introduction of a new hook: https://bugs.python.org/issue41962)
+        self._thread = threading.Thread(
+            name="Tornado selector",
+            daemon=True,
```
Would it be possible to avoid daemon thread?

minrk replied:
#86128 discusses why this is a daemon thread. I believe we need to call

```python
with loop._select_cond:
            loop._closing_selector = True
            loop._select_cond.notify()
        try:
            loop._waker_w.send(b"a")
        except BlockingIOError:
            pass
```

prior to `thread.join()` for it to wake and exit properly. It looks like I can make it a non-daemon thread if I register the atexit hook with the private `threading._register_atexit`.

I believe this is only needed to handle process teardown for unclosed event loops. If we can strictly guarantee that `EventLoop.close()` is called before non-daemon threads are joined, then I don't think this is an issue. But I don't personally know a reliable way other than `threading._register_atexit`.

`atexit.register` only runs _after_ threads are joined, not non-daemon threads (hence #86128), meaning that switching the thread to non-daemon will hang process exit until `select` returns, because no wake is called. A public hook that does exactly what `threading._register_atexit` does would be great. But since this is now in the stdlib, I could use that here, I suppose (#86128 remains a problem for non-stdlib code, of course).

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,298 @@
+            callback = cb_map[fd]
+        except KeyError:
+            return
+        callback()
```
What happens in case of exceptions? Can you call loop.call_exception_handler()?

minrk replied:
This method is exclusively scheduled via `call_soon_threadsafe`, so I think that happens already? You get this:

```
> python3 test.py
Exception in callback SelectorThread._handle_select([8], [6])
handle: <Handle SelectorThread._handle_select([8], [6])>
Traceback (most recent call last):
  File "/Users/minrk/conda/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/minrk/dev/py/cpython/Lib/asyncio/_selector_thread.py", line 232, in _handle_select
    self._handle_event(w, self._writers)
  File "/Users/minrk/dev/py/cpython/Lib/asyncio/_selector_thread.py", line 244, in _handle_event
    callback()
  File "/Users/minrk/dev/py/cpython/test.py", line 49, in write
    1/0
    ~^~
```

And if I call `loop.set_exception_handler(print)`, the output changes to:

```
<TestThreadEventLoop running=True closed=False debug=False> {'message': 'Exception in callback SelectorThread._handle_select([8], [6])', 'exception': ZeroDivisionError('division by zero'), 'handle': <Handle SelectorThread._handle_select([8], [6])>}
```

vstinner replied:
In case of exception, does _handle_select() continue calling following callbacks? I don't think so, and that's my main concern.

vstinner commented on Lib/asyncio/proactor_events.py:
```diff
@@ -641,6 +643,17 @@ def __init__(self, proactor):
+    def _get_selector_thread(self):
+        """Return the SelectorThread.
+
+        creating it on first request,
```
```suggestion
        Creating it on first request,
```

minrk committed:
check and split fd-like to int FD in proactor add_reader

matches tornado behavior

tornado handles this in IOLoop, so we need to include it

minrk committed:
Address revieww

- remove some older Python compatibility, since this is in the stdlib now
- resolve race between selector thread call soon and EventLoop.close

minrk committed:
asyncio._selector_thread: avoid daemon thread

rely on private threading._register_atexit to run cleanup prior to thread join

minrk committed:
add some docstrings to SelectorThread methods

based on confusion in feedback

minrk committed:
use versionchanged:: next

according to devguide

vstinner commented on Doc/library/asyncio-platforms.rst:
```diff
@@ -59,15 +59,20 @@ All event loops on Windows do not support the following methods:
+  When called, :func:`select.select` is run in an additional thread.
 
 The resolution of the monotonic clock on Windows is usually around 15.6
 milliseconds.  The best resolution is 0.5 milliseconds. The resolution depends on the
```
"The resolution of the monotonic clock on Windows is usually around 15.6 milliseconds."

Note unrelated to this PR: that's [no longer true in Python 3.13](https://docs.python.org/dev/whatsnew/3.13.html):

> On Windows, monotonic() now uses the QueryPerformanceCounter() clock for a resolution of 1 microsecond, instead of the GetTickCount64() clock which has a resolution of 15.6 milliseconds.

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,312 @@
+                    return
+            self._real_loop.call_soon_threadsafe(
+                    self._handle_select, rs, ws, context=self._main_thread_ctx
+                )
```
If the event loop is slow to process call_soon_threadsafe() queue, many _handle_select() callbacks will be queued, no?

For example, if a socket received data, select.select() returns immediately that the socket is ready for reading. _handle_select() callback is queud... and select.select() is called again, returns immediately, etc. No?

Or does _select_cond.wait() above and self._start_select() in _handle_select() prevent this issue?

minrk replied:
The Condition prevents this - it won't call `select` again until `_start_select` is called from the event loop. I'll try to add a test that verifies that's true.

minrk replied:
Added test that verifies this.

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,312 @@
+            self._handle_event(r, self._readers)
+        for w in ws:
+            self._handle_event(w, self._writers)
+        self._start_select()
```
Is _start_select() called if _handle_event() raises an exception?

minrk replied:
It doesn't. I'm going to wrap `_handle_event` to call `handle._run()` to ensure everything's wrapped appropriately.

minrk replied:
I switched this to use `events.Handle`, like SelectorEventLoop's `add_reader`, and added tests to ensure exceptions in callbacks don't cause problems.

vstinner replied:
Thanks, using Handle seems to be the right fix to make the code more reliable when exceptions occur.

minrk committed:
use Handles to ensure selector thread don't raise and halt handling

minrk committed:
more robust selector thread test

for thread asynciness

vstinner commented on Lib/test/test_asyncio/test_windows_events.py:
```diff
@@ -323,6 +323,59 @@ def threadMain():
+            self.loop.add_reader(b, read)
+            _selector_thread = self.loop._selector_thread
+            assert b.fileno() in _selector_thread._readers
+            assert _selector_thread is not None
```
Please use unittest methods, such as: assertIn(a, b) and assertIsNone(a). Same remark for other checks.

minrk committed:
use assert methods

vstinner commented on Lib/test/test_asyncio/test_selector_thread.py:
```diff
@@ -0,0 +1,123 @@
+            self.selector_thread.add_reader(b, mock_recv)
+            # ready event, but main event loop is blocked for some time
+            time.sleep(0.1)
+            recvd = await asyncio.wait_for(first_recv, timeout=10)
```
```suggestion
            recvd = await asyncio.wait_for(first_recv, timeout=support.SHORT_TIMEOUT)
```

You need to add `from test import support` at the top.

vstinner commented on Lib/test/test_asyncio/test_selector_thread.py:
```diff
@@ -0,0 +1,123 @@
+from unittest import mock
+
+
+class SelectorThreadTest((unittest.IsolatedAsyncioTestCase)):
```
```suggestion
class SelectorThreadTest(unittest.IsolatedAsyncioTestCase):
```

vstinner commented on Lib/test/test_asyncio/test_selector_thread.py:
```diff
@@ -0,0 +1,123 @@
+        with mock.patch.object(
+            selector_thread, "_start_select", wraps=selector_thread._start_select
+        ) as start_select:
+            await asyncio.wait_for(bad_recv_done, timeout=10)
```
```suggestion
            await asyncio.wait_for(bad_recv_done, timeout=support.SHORT_TIMEOUT)
```

vstinner commented on Lib/test/test_asyncio/test_windows_events.py:
```diff
@@ -323,6 +323,59 @@ def threadMain():
+            self.loop.add_writer(a, write)
+            self.assertIs(self.loop._selector_thread, _selector_thread)
+            self.assertIn(a.fileno(), _selector_thread._writers)
+            msg = await asyncio.wait_for(read_future, timeout=10)
```
```suggestion
            msg = await asyncio.wait_for(read_future, timeout=support.SHORT_TIMEOUT)
```

vstinner commented on Lib/asyncio/_selector_thread.py:
```diff
@@ -0,0 +1,323 @@
+        """Return fd, file object
+
+        Keeps a handle on the fileobject given,
+        but always registers integer FD
```
```suggestion
        but always registers integer FD.
```

minrk committed:
Apply suggestions from code review

Co-authored-by: Victor Stinner <vstinner@python.org>

minrk committed:
from test import support

minrk committed:
apply suggestions from review

Co-authored-by: Victor Stinner <vstinner@python.org>

minrk committed:
copy event loop policy teardown from other asyncio tests

vstinner commented on Lib/test/test_asyncio/test_selector_thread.py:
```diff
@@ -0,0 +1,128 @@
+        with mock.patch("select.select", wraps=select.select) as mock_select:
+            self.selector_thread.add_reader(b, mock_recv)
+            # ready event, but main event loop is blocked for some time
+            time.sleep(0.1)
```
Is this sleep really needed? If yes, can it be replaced with a synchronization primitive instead of a sleep?

minrk replied:
I would love help writing tests to verify that things _don't_ happen in multithreaded code.

This was added to demonstrate an answer to the question in review that the select loop doesn't keep firing away if the main loop is slow to process one event, and was the best I could come up with to demonstrate that it really is waiting and only calls select once.

The sequence of events that we want to wait for is that the select thread is waiting in select_cond.wait() after select is called the first time. Maybe there's a way to instrument that.

vstinner reviewed (approved):
Overall, the change *looks* good to me. I didn't test it. I just trust tests and the code that I read :-) I would prefer to have a second review from another core dev. 

@kumaraditya303: Would you be interested to review this change?

vstinner commented on Lib/test/test_asyncio/test_selector_thread.py:
```diff
@@ -0,0 +1,128 @@
+
+
+def tearDownModule():
+    asyncio.events._set_event_loop_policy(None)
```
It seems like asyncio.events._set_event_loop_policy() has been removed.

minrk replied:
thanks, updated and fixed lint
